src/evm/chandrupatla.rs

```diff
@@ -0,0 +1,1027 @@
+//! - SciPy: <https://github.com/scipy/scipy/blob/v1.15.2/scipy/optimize/_chandrupatla.py>
+//!
+//! ## Algorithm
+//!
+//! The algorithm maintains three points:
+//! - `x1`: One endpoint of the bracket (f1 and f2 have opposite signs)
+//! - `x2`: Other endpoint, also the current best estimate
+//! - `x3`: Previous value of the contrapoint (used for IQI)
+//!
+//! At each iteration:
+//! 1. Compute ξ (xi) = position ratio and φ (phi) = function value ratio
+//! 2. Check if IQI is favorable: `(1 - √(1-ξ)) < φ < √ξ`
+//! 3. If favorable, use IQI to estimate the next point
+//! 4. Otherwise, use bisection (t = 0.5)
+
+use num_bigint::BigUint;
+use num_traits::{FromPrimitive, ToPrimitive};
+use tycho_common::{
+    models::token::Token,
+    simulation::{errors::SimulationError, protocol_sim::ProtocolSim},
+};
+
+// =============================================================================
+// Configuration
+// =============================================================================
+
+/// Configuration parameters for Chandrupatla's root-finding algorithm.
+///
+/// These parameters control convergence behavior, numerical stability,
+/// and iteration limits.
+#[derive(Debug, Clone, Copy)]
+pub struct ChandrupatlaConfig {
+    /// Relative tolerance for price convergence.
+    /// The algorithm converges when `|actual - target| / target < tolerance`.
+    /// Default: 0.00001 (0.001%)
+    pub tolerance: f64,
+
+    /// Maximum number of iterations before giving up.
+    /// Default: 100
+    pub max_iterations: u32,
+
+    /// Minimum divisor to avoid division by zero in numerical computations.
+    /// Default: 1e-12
+    pub min_divisor: f64,
+
+    /// Minimum value for the interpolation parameter t.
+    /// Keeps the next estimate away from bracket endpoints.
+    /// The parameter t is clamped to [t_min, 1 - t_min].
+    /// Default: 0.05
+    pub t_min: f64,
+}
+
+impl Default for ChandrupatlaConfig {
+    fn default() -> Self {
+        Self {
+            tolerance: 0.00001,
+            max_iterations: 100,
+            min_divisor: 1e-12,
+            t_min: 0.05,
+        }
+    }
+}
+
+impl ChandrupatlaConfig {
+    /// Creates a new configuration with default values.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Sets the convergence tolerance.
+    pub fn with_tolerance(mut self, tolerance: f64) -> Self {
+        self.tolerance = tolerance;
+        self
+    }
+
+    /// Sets the maximum number of iterations.
+    pub fn with_max_iterations(mut self, max_iterations: u32) -> Self {
+        self.max_iterations = max_iterations;
+        self
+    }
+
+    /// Sets the minimum divisor for numerical stability.
+    pub fn with_min_divisor(mut self, min_divisor: f64) -> Self {
+        self.min_divisor = min_divisor;
+        self
+    }
+
+    /// Sets the minimum interpolation parameter.
+    pub fn with_t_min(mut self, t_min: f64) -> Self {
+        self.t_min = t_min;
+        self
+    }
+}
```

Comment:
why builder pattern? just 4 parameters.. One new() function with the 4 is enough IMO